cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -43,173 +54,199 @@
-      throws IOException, DatasetManagementException {
+    protected void processNotifications(DatasetContext context, Iterator<Notification> notifications) {
+      AppMetadataStore appMetadataStore = getAppMetadataStore(context);
+      while (notifications.hasNext()) {
+        processNotification(appMetadataStore, notifications.next());
+      }
+    }
+
+    private void processNotification(AppMetadataStore appMetadataStore, Notification notification) {
       Map<String, String> properties = notification.getProperties();
       // Required parameters
       String programRun = properties.get(ProgramOptionConstants.PROGRAM_RUN_ID);
       String programStatus = properties.get(ProgramOptionConstants.PROGRAM_STATUS);
 
-      // Ignore notifications which specify an invalid ProgramRunId or ProgramRunStatus
+      // Ignore notifications which specify an invalid ProgramRunId or ProgramRunStatus, which shouldn't happen
       if (programRun == null || programStatus == null) {
+        SAMPLING_LOG.warn("Ignore notification that misses program run state information, {}", notification);
         return;
       }
 
       ProgramRunStatus programRunStatus;
       try {
         programRunStatus = ProgramRunStatus.valueOf(programStatus);
       } catch (IllegalArgumentException e) {
-        LOG.warn("Invalid program run status {} passed in notification for program {}",
-                 programStatus, programRun);
+        LOG.warn("Ignore notification with invalid program run status {} for program {}, {}",
+                 programStatus, programRun, notification);
         return;
       }
 
       ProgramRunId programRunId = GSON.fromJson(programRun, ProgramRunId.class);
       ProgramId programId = programRunId.getParent();
 
-      ApplicationMeta meta = getAppMetadataStore(context).getApplication(programRunId.getNamespace(),
-                                                                         programRunId.getApplication(),
-                                                                         programRunId.getVersion());
+      ApplicationMeta meta = appMetadataStore.getApplication(programRunId.getNamespace(),
+                                                             programRunId.getApplication(),
+                                                             programRunId.getVersion());
       // Check if the application exists
       if (meta == null) {
-        LOG.warn(new NotFoundException(programId.getParent()).getMessage());
+        LOG.warn("Ignore notification for application {} that doesn't exist, {}", programRunId, notification);
         return;
       }
       // Check if the program exists
       if (getProgramSpecFromApp(meta.getSpec(), programRunId) == null) {
-        LOG.warn(new NotFoundException(programId).getMessage());
+        LOG.warn("Ignore notification for program {} that doesn't exist, {}", programRunId, notification);
```

Comment:
Just looked at the `AppMetaStore` again, and it actually doesn't check, so the check has to happen here.